### DISCUSSION TITLE
Hook event name inconsistencies

### DISCUSSION THREAD

**Discussion by nickrum** [directus/directus#7744] [Feature Requests]
There are a few inconsistencies around hook event names. Let's discuss if we should change those before the v9 release. The current names can be found in the [docs](https://docs.directus.io/guides/api-hooks/#event-format-options).

#### `routes.init`, `routes.custom.init` and `middlewares.init` have `before` / `after` actions breaking the scope/action/before model. For all other events you can append `before` or omit it.

One option would be to simply remove the `after` action.

Alternatively, we could also add `after` as an optional alias for "omitting `before`" to all other events. I think there was at least some confusion in the past about why there is no `after` when there is a `before` (this could also be mitigated by better documentation, event name validation and Typescript hints in the future).

#### The `cron(* * * * *)` syntax doesn't play nicely with the scope/action/before model. Also, IMO _cron_ is just an implementation detail that might not immediately convey the purpose to someone not familiar with cron.

I'd suggest using the `<scope>.<action>` syntax here as well. We could go with `interval.* * * * *` (the name _interval_ is already used to describe the hook in the [docs](https://docs.directus.io/guides/api-hooks/#interval-cron)) or `schedule.* * * * *`. Though, the syntax might actually be more confusing this way.

#### `oauth.:provider` is the only event name with an additional "parameter", making the distinction between scope, action and before less clear.

Maybe we could come up with a new syntax for scopes with a "sub-scope" using a different delimiter.

#### The `init` scope doesn't really fit with the other scopes.

Maybe we could make it an action under the `server` scope alongside `start` and `stop`.

##### I tried to sort these inconsistencies by decreasing severity.

**Comment by rijkvanzanten** [comment]
One thing to keep in mind on the delimiter question is that the event emitter we use supports wildcards based on that single configured delimiter, so you could do something like `oauth.*` to attach to every provider. We'd lose that functionality when introducing new delimiters for scope/subscope etc.

> There are a few inconsistencies around hook event names.

In general, I think it might prove difficult/impossible to strictly adhere to `<scope>.<action>.<before>` for every hook, as some of these hooks are just very different in what they aim to achieve. For the "item modification hooks" (eg `users.create` / `items.update.before` etc) the scope makes sense, but for things like `server.init` or `cron(* * * * *)` not so much.

> Also, IMO cron is just an implementation detail that might not immediately convey the purpose to someone not familiar with cron.

This is very true, _however_, the syntax used is "the CRON syntax", so if you wanna google what it does and how it works, the name `cron` is absolutely crucial in that context IMO. 

> I'd suggest using the <scope>.<action> syntax here as well.

I don't think that's going to fly. The delimiter with the `*` indicates a wildcard, which will break for the cron thing. I also see the "cron configuration" as a parameter of the cron usage, not a different scope 🤔

**Reply by nickrum** [reply]
> One thing to keep in mind on the delimiter question is that the event emitter we use supports wildcards based on that single configured delimiter, so you could do something like `oauth.*` to attach to every provider. We'd lose that functionality when introducing new delimiters for scope/subscope etc.

I noticed that as well after creating the discussion. Let's keep `oauth` as is then 👍 After all, subscopes _are_ scopes as well, so this might work fine with the current model.

> ... but for things like `server.init` ...

Is that a "yes" for renaming `init` to `server.init`? 😁

> This is very true, _however_, the syntax used is "the CRON syntax", so if you wanna google what it does and how it works, the name `cron` is absolutely crucial in that context IMO.

I agree with that, _but_, when I am looking for a way to schedule a certain action or run it in certain intervals, I would look for something that says "interval" or "schedule". Only after that I would expect to learn about the details on how to schedule my action and to learn about the cron syntax.

> I don't think that's going to fly. The delimiter with the `*` indicates a wildcard, which will break for the cron thing. I also see the "cron configuration" as a parameter of the cron usage, not a different scope 🤔

Well, it is in fact a wildcard, just not in terms of eventemitter2. The cron thing could also be seen as an action on the cron scope.

**Reply by benhaynes** [reply]
I understand that the format is `cron`, but I'd be open to renaming to something like `schedule` so it's a bit more obvious what it does. We can then include the Cron syntax for fixed times, dates, or intervals in the docs.

**Reply by rijkvanzanten** [reply]
> I agree with that, but, when I am looking for a way to schedule a certain action or run it in certain intervals, I would look for something that says "interval" or "schedule".

True! But that's a docs-problem, not a hooks naming problem 🙂 

To me, if you see `cron(* * * * *)` in the wild, it'll be easier to understand what's going on than when you'd see `schedule(* * * * *)` as cron is universal and easily searchable on the web, where schedule would require directus specific documentation to learn more about 🤔

**Comment by smilledge** [comment]
Another solution could be to pass functions for listening to events and scheduling jobs?

```typescript
interface HookAPI {
	listen: (eventName: string, callback: () => any) => void;
	schedule: (expression: string, callback: () => any) => void;
}

interface HookContext {
	env: Record<string, string>;
	database: Knex;
	// etc...
}

export default function registerHook(hookApi: HookAPI, context: HookContext) {
	const { listen, schedule } = hookApi;

	listen('middlewares.init', () => {
		// ✨️
	});
	
	schedule('* * * * *', () => {
		// ✨️
	});
}
```

This would solve the cron issues and might be more future proof as you can extend the `HookAPI` without breaking backwards comparability.

**Comment by nickrum** [comment]
This has been addressed before the v9 release!
